linkerd/stack/src/failfast/gate.rs

```diff
@@ -0,0 +1,102 @@
+use super::Shared;
+use futures::{ready, FutureExt};
+use std::{
+    sync::{atomic::Ordering, Arc},
+    task::{Context, Poll},
+};
+use tokio_util::sync::ReusableBoxFuture;
+use tracing::trace;
+
+/// A middleware which, when paired with a [`FailFast`] middleware, advertises
+/// the *actual* readiness state of the [`FailFast`]'s inner service up the
+/// stack.
+///
+/// A [`FailFast`]/[`Gate`] pair is primarily intended to be used in
+/// conjunction with a `tower::Buffer`. By placing the [`FailFast`] middleware
+/// inside of the `Buffer` and the `Gate` middleware outside of the buffer,
+/// the buffer's queue can be proactively drained when the inner service enters
+/// failfast, while the outer `Gate` middleware will continue to return
+/// [`Poll::Pending`] from its `poll_ready` method. This can be used to fail any
+/// requests that have already been dispatched to the inner service while it is in
+/// failfast, while allowing a load balancer or other traffic distributor to
+/// send any new requests to a different backend until this backend actually
+/// becomes available.
+///
+/// A `Layer`, such as a `Buffer` layer, may be wrapped in a new `Layer` which
+/// produces a [`FailFast`]/[`Gate`] pair around the inner `Layer`'s
+/// service using the [`FailFast::wrap_layer`] function.
```

Comment:
moving this to its own module broke the link references to failfast
```suggestion
/// service using the [`FailFast::layer_gated`] function.
///
/// [`FailFast`]: super::FailFast
/// [`FailFast::layer_gated`]: super::FailFast::layer_gated
```